### PR TITLE
Add accurate hash join size functions

### PR DESCRIPTION
**Pull request by PointKernel**
Addresses https://github.com/rapidsai/cudf/issues/8237

This PR adds 3 join size APIs (`hash_join::inner_join_size`, `hash_join::left_join_size` and `hash_join::full_join_size`) into `hash_join` class, one for each type of join that returns the exact number of matches with the specified probe table. It completely removed the deprecated size estimation logic in the current implementation.

Also, this PR updates the existing join APIs by adding an optional `output_size` as an argument. If `output_size.has_value()`, we take that value directly for further computation. Otherwise, the target join will internally invoke its corresponding size function.

`TODO`: the current `full_join_size` uses a 2-step algorithm similar to what's used in `hash_join::full_join`. It duplicates certain computations with `full_join` also thus should be refactored during `cuco` integration.

### CODE REVIEW

File: cpp/src/join/hash_join.cu [hunk 1]
```diff
@@ -226,39 +230,59 @@ probe_join_hash_table(cudf::table_device_view build_table,
...
+    join_size = output_size.value();
+  }
+  // Otherwise, compute the exact output size
+  else {
+    constexpr cudf::detail::join_kind ProbeJoinKind =
+      (JoinKind == cudf::detail::join_kind::FULL_JOIN) ? cudf::detail::join_kind::LEFT_JOIN
+                                                       : JoinKind;
+    join_size = compute_join_output_size<ProbeJoinKind>(
+      build_table, probe_table, hash_table, compare_nulls, stream);
+  }
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:249]:
```suggestion 
    constexpr cudf::detail::join_kind ProbeJoinKind =
      (JoinKind == cudf::detail::join_kind::FULL_JOIN) ? cudf::detail::join_kind::LEFT_JOIN
                                                       : JoinKind;
                                                       
  std::size_t  const join_size =  output_size.value_or( compute_join_output_size<ProbeJoinKind>(
      build_table, probe_table, hash_table, compare_nulls, stream);
```

File: cpp/include/cudf/join.hpp [hunk 2]
```diff
@@ -579,9 +588,57 @@ class hash_join {
...
-            null_equality compare_nulls         = null_equality::EQUAL,
-            rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
-            rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource()) const;
+            null_equality compare_nulls            = null_equality::EQUAL,
+            std::optional<std::size_t> output_size = {},
+            rmm::cuda_stream_view stream           = rmm::cuda_stream_default,
+            rmm::mr::device_memory_resource* mr    = rmm::mr::get_current_device_resource()) const;
+
+  /**
+   * Returns the exact number of output when performing an inner join with the specified probe
```
Comment (davidwendt) [inline cpp/include/cudf/join.hpp:597]: What unit does `exact number of output` reference? Is it number of rows or perhaps number of bytes?

Comment (PointKernel) [reply cpp/include/cudf/join.hpp:597]: It should be number of rows. I will update the doc accordingly.

**Review by jrhemstad — APPROVED** [review]

File: cpp/src/join/hash_join.cu [hunk 3]
```diff
@@ -323,40 +423,95 @@ std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
...
+std::size_t hash_join::hash_join_impl::full_join_size(cudf::table_view const &probe,
+                                                      null_equality compare_nulls,
+                                                      rmm::cuda_stream_view stream,
+                                                      rmm::mr::device_memory_resource *mr) const
+{
+  CUDF_FUNC_RANGE();
+
+  // Trivial left join case - exit early
+  if (!_hash_table) { return probe.num_rows(); }
+  CUDF_EXPECTS(_hash_table, "Hash table of hash join is null.");
```
Comment (robertmaynard) [inline cpp/src/join/hash_join.cu:501]: I think that this `CUDF_EXPECTS` can be removed, the `!_hash_table` check one line above will mean only valid hash tables will be around

File: cpp/src/join/hash_join.cu [hunk 4]
```diff
@@ -323,40 +423,95 @@ std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
...
+
+std::size_t hash_join::hash_join_impl::left_join_size(cudf::table_view const &probe,
+                                                      null_equality compare_nulls,
+                                                      rmm::cuda_stream_view stream) const
+{
+  CUDF_FUNC_RANGE();
+
+  // Trivial left join case - exit early
+  if (!_hash_table) { return probe.num_rows(); }
+  CUDF_EXPECTS(_hash_table, "Hash table of hash join is null.");
```
Comment (robertmaynard) [inline cpp/src/join/hash_join.cu:483]: I think that this `CUDF_EXPECTS` can be removed, the `!_hash_table` check one line above will mean only valid hash tables will be around

File: cpp/src/join/hash_join.cu (hunk 3 above)
Comment (PointKernel) [reply cpp/src/join/hash_join.cu:501]: Good point.

**Review by robertmaynard — APPROVED** [review]

**Review by revans2 — APPROVED** [review]
The API looks great for what we would want.

**Comment by PointKernel** [comment]
@gpucibot merge

**Pull request merged**